File: app/src/main/java/com/todoroo/astrid/api/CaldavFilter.java
package com.todoroo.astrid.api;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.todoroo.andlib.sql.Criterion;
import com.todoroo.andlib.sql.Join;
import com.todoroo.andlib.sql.QueryTemplate;
import com.todoroo.astrid.data.Task;

import org.tasks.R;
import org.tasks.data.CaldavCalendar;
import org.tasks.data.CaldavTask;
import org.tasks.data.TaskDao;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class CaldavFilter extends Filter {

  /** Parcelable Creator Object */
  public static final Parcelable.Creator<CaldavFilter> CREATOR =
      new Parcelable.Creator<>() {

        /** {@inheritDoc} */
        @Override
        public CaldavFilter createFromParcel(Parcel source) {
          CaldavFilter item = new CaldavFilter();
          item.readFromParcel(source);
          return item;
        }

        /** {@inheritDoc} */
        @Override
        public CaldavFilter[] newArray(int size) {
          return new CaldavFilter[size];
        }
      };

  private CaldavCalendar calendar;

  private CaldavFilter() {
    super();
  }

  public CaldavFilter(CaldavCalendar calendar) {
    super(calendar.getName(), queryTemplate(calendar), getValuesForNewTask(calendar));
    this.calendar = calendar;
    id = calendar.getId();
    tint = calendar.getColor();
    icon = calendar.getIcon();
    order = calendar.getOrder();
  }

  private static QueryTemplate queryTemplate(CaldavCalendar caldavCalendar) {
    return new QueryTemplate()
        .join(Join.left(CaldavTask.TABLE, Task.ID.eq(CaldavTask.TASK)))
        .where(getCriterion(caldavCalendar));
  }

  private static Criterion getCriterion(CaldavCalendar caldavCalendar) {
    return Criterion.and(
        TaskDao.TaskCriteria.activeAndVisible(),
        CaldavTask.DELETED.eq(0),
        CaldavTask.CALENDAR.eq(caldavCalendar.getUuid()));
  }

  private static Map<String, Object> getValuesForNewTask(CaldavCalendar caldavCalendar) {
    Map<String, Object> result = new HashMap<>();
    result.put(CaldavTask.KEY, caldavCalendar.getUuid());
    return result;
  }

  public String getUuid() {
    return calendar.getUuid();
  }

  public String getAccount() {
    return calendar.getAccount();
  }

  public CaldavCalendar getCalendar() {
    return calendar;
  }

  @Override
  public boolean isReadOnly() {
    return calendar.getAccess() == CaldavCalendar.ACCESS_READ_ONLY;
  }

  /** {@inheritDoc} */
  @Override
  public void writeToParcel(Parcel dest, int flags) {
    super.writeToParcel(dest, flags);
    dest.writeParcelable(calendar, 0);
  }

  @Override
  protected void readFromParcel(Parcel source) {
    super.readFromParcel(source);
    calendar = source.readParcelable(getClass().getClassLoader());
  }

  @Override
  public boolean supportsManualSort() {
    return true;
  }

  @Override
  public int getMenu() {
    return R.menu.menu_caldav_list_fragment;
  }

  @Override
  public boolean areContentsTheSame(@NonNull FilterListItem other) {
    return super.areContentsTheSame(other)
        && Objects.equals(calendar, ((CaldavFilter) other).calendar);
  }
}

File: app/src/main/java/com/todoroo/astrid/api/CaldavFilter.kt
package com.todoroo.astrid.api

import org.tasks.data.sql.Criterion.Companion.and
import org.tasks.data.sql.Join.Companion.left
import org.tasks.data.sql.QueryTemplate
import com.todoroo.andlib.utility.AndroidUtilities
import kotlinx.parcelize.Parcelize
import org.tasks.data.entity.CaldavCalendar
import org.tasks.data.entity.CaldavTask
import org.tasks.data.NO_COUNT
import org.tasks.data.entity.Task
import org.tasks.data.dao.TaskDao.TaskCriteria.activeAndVisible
import org.tasks.filters.Filter
import org.tasks.filters.FilterListItem
import org.tasks.filters.mapToSerializedString

@Parcelize
data class CaldavFilter(
    val calendar: CaldavCalendar,
    val principals: Int = 0,
    override val count: Int = NO_COUNT,
) : Filter {
    override val title: String?
        get() = calendar.name
    override val sql: String
        get() = QueryTemplate()
            .join(left(CaldavTask.TABLE, Task.ID.eq(CaldavTask.TASK)))
            .where(
                and(
                    activeAndVisible(),
                    CaldavTask.DELETED.eq(0),
                    CaldavTask.CALENDAR.eq(calendar.uuid)
                )
            )
            .toString()
    override val valuesForNewTasks: String
        get() = mapToSerializedString(mapOf(CaldavTask.KEY to calendar.uuid!!))

    override val order: Int
        get() = calendar.order

    override val tint: Int
        get() = calendar.color
    override val icon: Int
        get() = calendar.getIcon()!!

    val uuid: String
        get() = calendar.uuid!!
    val account: String
        get() = calendar.account!!

    override val isReadOnly: Boolean
        get() = calendar.access == CaldavCalendar.ACCESS_READ_ONLY

    override fun supportsManualSort() = true

    override fun areItemsTheSame(other: FilterListItem): Boolean {
        return other is CaldavFilter && calendar.id == other.calendar.id
    }
}
